extensions/functions_string.yaml

```diff
@@ -163,3 +163,355 @@ scalar_functions:
+        return: i8
+      - args:
+          - value: "varchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "varchar<L1>"
+            name: "substring"
+            description: The substring to search for.
+        return: i8
+      - args:
+          - value: "varchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "string"
+            name: "substring"
+            description: The substring to search for.
+        return: i8
+      - args:
+          - value: "varchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "fixedchar<L1>"
+            name: "substring"
+            description: The substring to search for.
+        return: i8
+      - args:
+          - value: "fixedchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "fixedchar<L1>"
+            name: "substring"
+            description: The substring to search for.
+        return: i8
+      - args:
+          - value: "fixedchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "string"
+            name: "substring"
+            description: The substring to search for.
+        return: i8
+      - args:
+          - value: "fixedchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "varchar<L1>"
+            name: "substring"
+            description: The substring to search for.
+        return: i8
+  -
+    name: regexp_strpos
+    description: >-
+      Return the position of the first occurrence of a regular expression pattern in another
+      string. The first character of the string is at position 1. If no occurrence is found,
+      0 is returned.
+    impls:
+      - args:
+          - options: [ CASE_SENSITIVE, CASE_INSENSITIVE ]
+            required: false
+          - value: "string"
+            name: "input"
+            description: The input string.
+          - value: "string"
+            name: "pattern"
+            description: The regular expression pattern.
+      - args:
+          - options: [ CASE_SENSITIVE, CASE_INSENSITIVE ]
+            required: false
+          - value: "varchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "varchar<L1>"
+            name: "pattern"
+            description: The regular expression pattern.
+        return: i8
+      - args:
+          - options: [ CASE_SENSITIVE, CASE_INSENSITIVE ]
+            required: false
+          - value: "fixedchar<L1>"
+            name: "input"
+            description: The input string.
+          - value: "fixedchar<L1>"
+            name: "pattern"
+            description: The regular expression pattern.
+        return: i8
+  -
+    name: count_substring
+    description: Returns the number of times a substring occurs in an input string.
```

Comment:
Good question.  I couldn't find if there was a standard for databases around this, but my first guess is that default would be non-overlapping, since that seems like a more common thing to do.  Would it make sense to have non-overlapping as default and provide an options for this function?

This was in the Acero code, but it doesn't specify about overlap either:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_string_ascii.cc#L1875-L1880

Reply:
I also figured it'd be without overlap, but it wouldn't hurt to specify. I'm fine with either just putting that in the description or adding an option for it.